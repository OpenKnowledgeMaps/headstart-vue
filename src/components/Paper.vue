<template>
    <g
            class="paper"
            :id="'paper' + id"
            :transform="translateString"
            style="display: block;"
            @mouseover="$emit('bubbleMouseEnter', bubbleId)"
            @mouseout="$emit('bubbleMouseLeave', bubbleId)"
    >

        <path
                id="region"
                :d="paperPath"
                class="unframed"
                style="fill-opacity: 1;">
        </path>
        <path
                class="dogear"
                :d="dogearPath"></path>

        <foreignObject
                id="article_metadata"
                class="paperContent"
                :width="w"
                :height="h"
                @mouseover="$emit('bubbleMouseEnter', bubbleId)"
                @mouseout="$emit('bubbleMouseLeave', bubbleId)"
        >
            <body>
                <div class="paper_holder">

                    <div class="metadata" :style="{ height: metadata_height + 'px', width: metadata_width + 'px'}">
                        <div id="icons">
                            <p id="open-access-logo">&#xf09c;</p>
                        </div>
                        <p id="title" class="highlightable">
                            {{ title }}<br />
                        </p>
                        <p id="details" class="highlightable">{{ authors_string }}</p>
                        <p id="in">in <span class="highlightable">{{ published_in }} <span class="pubyear">({{ year }})</span></span></p>
                    </div>
                    <div class="readers">
                        <p id="readers"><span id="num-readers">{{ readers }}</span><span class="readers_entity"> {{ base_unit }}</span></p>

                    </div>
                </div>
            </body>
        </foreignObject>
    </g>
</template>

<script>
  export default {
    name: "Paper",
    props: {
      x: Number,
      y: Number,
      w: Number,
      h: Number,
      id: Number,
      bubbleId: Number,
      metadata_height: Number,
      metadata_width: Number,
      title: String,
      authors_string: String,
      published_in: String,
      year: String,
      readers: String,
      base_unit: String
    },
    computed: {
      translateString: function() {
        return "translate(" + this.x + "," + this.y + ")";
      },
      paperPath: function() {
        return `M 0 0 h ${0.9*this.w} l ${0.1*this.w} ${0.1*this.h} v ${0.9*this.h} h ${-this.w} v ${-this.h}`
      },
      dogearPath: function() {
        return `M ${0.9*this.w} 0 v ${0.1*this.w} h ${0.1*this.h}`
      }
    }
  }
</script>

<style scoped>

</style>