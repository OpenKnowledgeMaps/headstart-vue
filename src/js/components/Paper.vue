<template src="../../templates/paper.html"></template>

<script>
  export default {
    name: "Paper",
    props: {
      x: Number,
      y: Number,
      w: Number,
      h: Number,
      id: Number,
      bubbleId: Number,
      metadata_height: Number,
      metadata_width: Number,
      title: String,
      authors_string: String,
      published_in: String,
      year: String,
      readers: String,
      base_unit: String
    },
    computed: {
      translateString: function() {
        return "translate(" + this.x + "," + this.y + ")";
      },
      paperPath: function() {
        return `M 0 0 h ${0.9*this.w} l ${0.1*this.w} ${0.1*this.h} v ${0.9*this.h} h ${-this.w} v ${-this.h}`
      },
      dogearPath: function() {
        return `M ${0.9*this.w} 0 v ${0.1*this.w} h ${0.1*this.h}`
      }
    }
  }
</script>

<style scoped>

</style>
