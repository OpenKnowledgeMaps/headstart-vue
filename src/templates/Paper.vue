<script src="../js/Paper.js">
</script>

<template>
<g
        class="paper"
        :id="'paper' + id"
        :transform="translateString"
        style="display: block;"
        @mouseover="$emit('mouseEnter', id)"
        @mouseout="$emit('mouseLeave', id)"
        @click="$emit('click', id)"
        @dblclick.prevent="$emit('doubleClick', id)"
>

    <path
            id="region"
            :d="paperPath"
            class="unframed"
            style="fill-opacity: 1;">
    </path>
    <path
            class="dogear"
            :d="dogearPath"></path>

    <foreignObject
            id="article_metadata"
            class="paperContent"
            :width="w"
            :height="h"
            @mouseover="$emit('mouseEnter', id)"
            @mouseout="$emit('mouseLeave', id)"
            @click="$emit('click', id)"
            @dblclick.prevent="$emit('doubleClick', id)"
    >
        <body>
            <div class="paper_holder">

                <div class="metadata" :style="{ height: metadata.height + 'px', width: metadata.width + 'px'}">
                    <div id="icons">
                        <p id="open-access-logo">&#xf09c;</p>
                    </div>
                    <p id="title" class="highlightable">
                        {{ metadata.title }}<br />
                    </p>
                    <p id="details" class="highlightable">{{ metadata.authors_string }}</p>
                    <p id="in">in <span class="highlightable">{{ metadata.published_in }}
                        <span class="pubyear">({{ metadata.year }})</span></span></p>
                </div>
                <div class="readers">
                    <p id="readers"><span id="num-readers">{{ metadata.readers }}</span>
                        <span class="readers_entity"> {{ metadata.base_unit }}</span></p>

                </div>
            </div>
        </body>
    </foreignObject>
</g>
</template>

<style scoped>
</style>
