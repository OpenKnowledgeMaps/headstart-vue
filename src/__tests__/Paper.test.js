import Paper from '../templates/Paper';
import { shallowMount } from '@vue/test-utils';

const factory = (values = {}) => {
  return shallowMount(Paper, {
    propsData: { ...values}
  })
};
let wrapper, gWrapper, foreignObjectWrapper;

beforeEach(() => {
  wrapper = factory({
    id: 0,
    x: 300,
    y: 300,
    w: 22.28,
    h: 29,
    order: 0,
    metadata: {
      width: 20,
      height: 23,
      title: "De­vel­op­ment of re­la­tional pro­cess­ing in hot and cool tasks.",
      authors_string: "Katie M Bunch, Glenda An­drews",
      published_in: "Developmental neuropsychology",
      year: "(2012)",
      readers: "2",
      base_unit: "citations"
    },
    bubbleId: 1
  });

  gWrapper = wrapper.find('g');
  foreignObjectWrapper = wrapper.find('foreignobject');
});

describe('Paper', () => {
  it('matches a previously made snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('s g element reacts to mouseover events', () => {
    gWrapper.trigger('mouseover');
    expect(wrapper.emitted('mouseEnter')).toBeTruthy();
  });

  it('s g element reacts to mouseout events', () => {
      gWrapper.trigger('mouseout');
      expect(wrapper.emitted('mouseLeave')).toBeTruthy();
  });

  it('s g element reacts to click events', () => {
      gWrapper.trigger('click');
      expect(wrapper.emitted('click')).toBeTruthy();
  });

  it('s g element reacts to doubleclick events', () => {
      gWrapper.trigger('dblclick');
      expect(wrapper.emitted('doubleClick')).toBeTruthy();
  });

  it('s foreignObject element reacts to mouseover events', () => {
      foreignObjectWrapper.trigger('mouseover');
      expect(wrapper.emitted('mouseEnter')).toBeTruthy();
  });

  it('s foreignObject element reacts to mouseout events', () => {
      foreignObjectWrapper.trigger('mouseout');
      expect(wrapper.emitted('mouseLeave')).toBeTruthy();
  });

  it('s foreignObject element reacts to click events', () => {
      foreignObjectWrapper.trigger('click');
      expect(wrapper.emitted('click')).toBeTruthy();
  });

  it('s foreignObject element reacts to doubleclick events', () => {
      foreignObjectWrapper.trigger('dblclick');
      expect(wrapper.emitted('doubleClick')).toBeTruthy();
  });
});
