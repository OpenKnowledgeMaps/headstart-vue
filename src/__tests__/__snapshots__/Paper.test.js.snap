// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paper matches a previously made snapshot 1`] = `
<g id="paper0" transform="translate(300,300)" class="paper" style="display: block;">
  <path id="region" d="M 0 0 h 20.052000000000003 l 2.228 2.9000000000000004 v 26.1 h -22.28 v -29" class="unframed" style="fill-opacity: 1;"></path>
  <path d="M 20.052000000000003 0 v 2.228 h 2.9000000000000004" class="dogear"></path>
  <foreignObject id="article_metadata" width="22.28" height="29" class="paperContent">

    <body>
      <div class="paper_holder">
        <div class="metadata" style="height: 23px; width: 20px;">
          <div id="icons">
            <p id="open-access-logo"></p>
          </div>
          <p id="title" class="highlightable">
            De­vel­op­ment of re­la­tional pro­cess­ing in hot and cool tasks.
            <br>
          </p>
          <p id="details" class="highlightable">Katie M Bunch, Glenda An­drews</p>
          <p id="in">in <span class="highlightable">Developmental neuropsychology 
                        <span class="pubyear">((2012))</span></span>
          </p>
        </div>
        <div class="readers">
          <p id="readers"><span id="num-readers">2</span> <span class="readers_entity"> citations</span></p>
        </div>
      </div>
    </body>
  </foreignObject>
</g>
`;
